### PR TITLE
`ClientCertificateCredential`: better exception message for unsupported file extensions

### LINKED ISSUE #4907: ClientCertificateCredetial: make better exception description for non-.pem files
We currently only support `.pem` files. We should try to add support for `.pfx` files, but this work item is about another matter (#4909).
It must not be uncommon for people to try to use the credential with `.pfx` files.
Currently, the exception will read: "Authentication error: Failed to read certificate private key."
We shoud check the extension and say sonlthing like ".pfx certificates are not supported, only the .pem cretificates are supported. Please convert your .pfx to .pem using the command 'openssl pkcs12 -in file.pfx -out file.pem -nodes'. "

### PR DESCRIPTION
**Pull request by antkmsft**
Closes #4907.

--
UWP CI is currently blocked until https://github.com/microsoft/vcpkg/issues/35279 is fixed / https://github.com/microsoft/vcpkg/pull/35116 is merged (#5181 would unblock)

### CODE REVIEW

**Review by RickWinter — APPROVED** [review]

File: sdk/identity/azure-identity/src/client_certificate_credential.cpp [hunk 1]
```diff
@@ -399,16 +400,33 @@ ClientCertificateCredential::ClientCertificateCredential(
     CertificateThumbprint mdVec;
     try
     {
+      if (clientCertificatePath.empty())
+      {
+        throw AuthenticationException("Certificate file path is empty.");
+      }
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/client_certificate_credential.cpp:406]: Why are we checking and throwing this within a try/catch? Shouldn't this input parameter validation go above?

File: sdk/identity/azure-identity/src/client_certificate_credential.cpp [hunk 2]
```diff
@@ -399,16 +400,33 @@ ClientCertificateCredential::ClientCertificateCredential(
...
+
+      using Azure::Core::_internal::StringExtensions;
+      std::string const PemExtension = ".pem";
+      auto const certFileExtensionStart = clientCertificatePath.find_last_of('.');
+      auto const certFileExtension = certFileExtensionStart != std::string::npos
+          ? clientCertificatePath.substr(certFileExtensionStart)
+          : std::string{};
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/client_certificate_credential.cpp:413]: This isn't a very resilient test for getting a file extension (what happens if the filename or path/directory name has a period in it). We can't use C++ 17, but maybe we can borrow some inspiration from [std::filesystem::path::extension](https://en.cppreference.com/w/cpp/filesystem/path/extension)

File: sdk/identity/azure-identity/src/client_certificate_credential.cpp [hunk 3]
```diff
@@ -399,16 +400,33 @@ ClientCertificateCredential::ClientCertificateCredential(
...
       std::tie(mdVec, m_pkey) = ReadPemCertificate(clientCertificatePath);
     }
-    catch (AuthenticationException&)
-    {
-      throw;
-    }
-    catch (std::exception& e)
+    catch (std::exception const& e)
     {
       // WIL does not throw AuthenticationException.
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/client_certificate_credential.cpp:427]: Unrelated to this PR, but what does WIL have to do with this implementation?

File: sdk/identity/azure-identity/test/ut/client_certificate_credential_test.cpp [hunk 4]
```diff
@@ -148,6 +148,77 @@ TEST_P(GetCredentialName, )
...
 
+TEST(ClientCertificateCredential, UnsupportedExtension)
+{
+  try
+  {
+    ClientCertificateCredential const cred(
+        "01234567-89ab-cdef-fedc-ba8976543210", "fedcba98-7654-3210-0123-456789abcdef", "file.pfx");
+
+    EXPECT_TRUE(
+        !"ClientCertificateCredential with unsupported extension (.pfx) is supposed to throw.");
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/ut/client_certificate_credential_test.cpp:159]: What does this `!` operator supposed to mean here on a string, when passed to `EXPECT_TRUE`?

Whatever the intent is here, I think we should write it clearly, so it's easier to understand.

File: sdk/identity/azure-identity/src/client_certificate_credential.cpp (hunk 1 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/client_certificate_credential.cpp:406]: Why not, and plus try-catch block is going to prepend `Identity: ClientCertificateCredential` to it.

File: sdk/identity/azure-identity/src/client_certificate_credential.cpp (hunk 3 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/client_certificate_credential.cpp:427]: I think that was added in a PR that added cert support via Win32 API, an it uses `wil`. I think the goal of the comment is to say that non-Azure-SDK code does throw all sorts of std::exeptions.

File: sdk/identity/azure-identity/src/client_certificate_credential.cpp (hunk 2 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/client_certificate_credential.cpp:413]: We search from the back of the string (`find_last_of()`), I think it is pretty resilient.
Feel free to log a work item to improve, maybe we need some facility in `Azure::Core` to provide the functionality similar to C++17 `std::filesystem`.

File: sdk/identity/azure-identity/test/ut/client_certificate_credential_test.cpp (hunk 4 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/client_certificate_credential_test.cpp:159]: This is a way to fail with message, it is kind of idiomatic - there is C and C++ code that uses it with `assert()`s, for example.
The way it works is that any string literal is never null, but `!` inverts that, i.e. `!""` is always false, and so when we `EXPECT_TRUE()` or `assert()`, the assertion always fails.

**Pull request merged**